src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -2498,6 +2499,337 @@ public static decimal Average<TSource>(this IEnumerable<TSource> source, Func<TS
         {
             return Enumerable.Average(Enumerable.Select(source, selector));
         }
+
+        public static TSource MaxBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector)
+        {
+            return source.MaxBy(keySelector, null);
+        }
```

Comment:
On the one hand you could further and check separately and `for` looped separately because the compiler optimises for loops on `T[]` differently to `IList<T>`. On the other, the increase cost in the extra check (and the increased amount of code to potentially have bugs) may not be worth it. I'm inclined to lean toward the simpler approach.